src/chkstat.cpp

```diff
@@ -21,1365 +21,1246 @@
+     * employ a backward compatibility logic here that prefers files in /usr
+     * but also recognizes files in /etc as a fallback.
+     */
+    const auto &config_root = m_config_root_path.getValue();
+
+    const auto usr_root = config_root + "/usr/share/permissions";
+    const auto etc_root = config_root + "/etc";
 
-    dir = opendir(directory);
-    if (dir)
+    // first add the central fixed permissions file
+    for (const auto &dir: {usr_root, etc_root})
     {
-        char** files = NULL;
-        size_t nfiles = 0;
-        struct dirent* d;
-        while ((d = readdir(dir)))
+        const auto path = dir + "/permissions";
+        if (existsFile(path))
         {
-            char* p;
-            if (!strcmp("..", d->d_name) || !strcmp(".", d->d_name))
-                continue;
+            m_profile_paths.push_back(path);
+            // only use the first one found
+            break;
+        }
+    }
 
-            /* filter out backup files */
-            if ((strlen(d->d_name)>2) && (d->d_name[strlen(d->d_name)-1] == '~'))
-                continue;
-            if (strstr(d->d_name,".rpmnew") || strstr(d->d_name,".rpmsave"))
-                continue;
+    // continue with predefined well-known profiles
+    for (const auto &profile: m_profiles)
+    {
+        if (!matchesAny(profile, PREDEFINED_PROFILES))
+            continue;
+
+        const auto base = std::string("/permissions.") + profile;
+
+        for (const auto &dir: {usr_root, etc_root})
+        {
+            std::string path = dir + base;
 
-            ensure_array((void**)&files, &nfiles);
-            if ((p = strchr(d->d_name, '.')))
+            if (existsFile(path))
             {
-                *p = '\0';
+                m_profile_paths.push_back(path);
+                // only use the first one found
+                break;
             }
-            files[nfiles++] = strdup(d->d_name);
         }
-        closedir(dir);
-        if (nfiles)
-        {
-            qsort(files, nfiles, sizeof(char*), compare);
-            for (i = 0; i < nfiles; ++i)
-            {
-                char fn[4096];
-                size_t l;
-                // skip duplicates
-                if (i && !strcmp(files[i-1], files[i]))
-                    continue;
-
-                snprintf(fn, sizeof(fn), "%s/%s", directory, files[i]);
-                if (access(fn, R_OK) == 0)
-                {
-                    ensure_array((void**)&permfiles, &npermfiles);
-                    permfiles[npermfiles++] = strdup(fn);
-                }
+    }
 
-                for (l = 0; l < nlevel; ++l)
-                {
-                    snprintf(fn, sizeof(fn), "%s/%s.%s", directory, files[i], level[l]);
+    // move on to package specific permissions
+    // these files are owned by the individual packages
+    // therefore while files are being moved from /etc to /usr we need to
+    // consider both locations, not only the first matching one like above.
```

Comment:
I'm not quite sure what you mean. We need to look in both locations until the migration is complete. So are you suggesting not to start the migration? Or are you afraid that package has files in both locations?

Reply:
For other files, we strictly prefer `/usr/share` over `/etc`. That comment explains that we treat the package-specific overrides differently. I'm questioning if that difference is really what we want. (And saying we are still free to change things since back-compat isn't an issue yet.)

Reply:
Since the package-specific files are owned by the package-specific RPMs it seems to me the most robust way to deal with it. Otherwise permissions for some packages would break in the meantime. Or packages would need to install symlinks for compatibility or something like that.

Reply:
I'm not suggesting we should parse only `/usr` and completely ignore `/etc` (or vice versa).

I just think we should/could use the same logic for package-specific files as we do for our own: only look in `/etc` if there's no file in `/usr`. Because there's no good reason that both should ever exist at the same time.

Reply:
So "Or are you afraid that a package has files in both locations?" applies as I asked you before. Wasn't quite on the same page with you.

Per-package files have no predefined names and can exist solely for specific profiles or as a base profile plus profile specific overrides.

So when we're afraid of strange states we could head into something like this:

```
/usr/share/permissions/permissions.d/postfix
/etc/permissions.d/postfix.paranoid
```

So basically when we find something in `/usr/share/permissions/permissions.d` we'd need to exclude any files `/etc/permissions.d/${base}*`. It will complicate the logic quite a bit and that is probably why it wasn't implemented in the first place.

I think I can make an attempt by simply carrying around a `std::set` with package basenames already seen. It would be cleaner in any case.

However how do we handle local overrides? Should `/etc/permissions.d` stay around after the migration is complete? Was it ever supported for Admins to place files in there like it is supported to modify `/etc/permissions.local`?

While we're already talking about this migration ... I don't like the path `/usr/share/permissions/permissions.d` very much. How about we change it to `/usr/share/permissions/packages.d`? Would make it more clear what it is for.

Reply:
> So basically when we find something in `/usr/share/permissions/permissions.d` we'd need to exclude any files `/etc/permissions.d/${base}*`.  It will complicate the logic quite a bit and that is probably why it wasn't implemented in the first place.

Yup, that's probably true. The more I think about it, the more I question how much sense it makes to spend too many cycles on 'broken' setups like this :)

> I think I can make an attempt by simply carrying around a `std::set` with package basenames already seen. It would be cleaner in any case.

Yes, that's the cleaner approach. Please don't feel obligated to do this in this PR though - you documenting the existing behavior just made me wonder if it's really what we want. We can  as well just make an issue to discuss this further.

> However how do we handle local overrides? Should `/etc/permissions.d` stay around after the migration is complete? Was it ever supported for Admins to place files in there like it is supported to modify `/etc/permissions.local`?

I don't think it's been decided yet. I don't think it was ever supposed to be supported, but it worked and all this stuff was so under-documented I don't think we can blame admins who ended up placing custom overrides there. On the other hand, since this stuff is so under-documented there can't be more than a hand-full of cases - so the question is how much we care about these.

> While we're already talking about this migration ... I don't like the path `/usr/share/permissions/permissions.d` very much. How about we change it to `/usr/share/permissions/packages.d`? Would make it more clear what it is for.

Yes, I don't like it much either, I just couldn't think of anything better back then. Maybe `package-overrides.d`? Just note that changing this needs careful synchronization with `rpmlint`.

Reply:
I have now implemented the approach using a `std::set`. It wasn't even that complex. I also added a test case to cover the new expectation.

Changing the name of the directory is a TODO, because it needs to be synced with rpmlint-checks.

If custom files exist for some users in `/etc/permissions.d` then they will still be processed as long as they don't clash with existing package basenames.